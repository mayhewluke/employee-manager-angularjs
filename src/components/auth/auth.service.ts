// tslint:disable-next-line:variable-name
export const logIn = (_email: string, _password: string) =>
  new Promise(() => null);
