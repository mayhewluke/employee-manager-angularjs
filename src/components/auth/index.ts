import angular from "angular";

import loginComponent from "./login.component";

export const authModule = angular.module("app.components.auth", [
  loginComponent,
]).name;

export * from "./auth.service";
