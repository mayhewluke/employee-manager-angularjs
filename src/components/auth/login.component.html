<form class="login-form form-horizontal" ng-submit="$ctrl.submit($ctrl.email, $ctrl.password)">
  <div class="form-group">
    <label class="col-xs-2 control-label">Email</label>
    <div class="col-xs-10">
      <input class="form-control" name="email" id="email" type="text" ng-model="$ctrl.email" required>
    </div>
  </div>

  <div class="form-group">
    <label class="col-xs-2 control-label">Password</label>
    <div class="col-xs-10">
      <input class="form-control" name="password" id="password" type="password" ng-model="$ctrl.password" required>
    </div>
  </div>

  <div class="form-group">
    <div class="col-xs-offset-2 col-xs-10">
      <button class="btn btn-primary" type="submit">Log in</button>
      <p class="error alert alert-danger" ng-if="$ctrl.hasError">Something went wrong!</p>
    </div>
  </div>
</form>
