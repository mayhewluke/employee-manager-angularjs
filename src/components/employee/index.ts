import angular from "angular";

import { employeeServiceModule } from "./employee.service";
import listComponent from "./list.component";

export const employeeModule = angular.module("app.components.employee", [
  employeeServiceModule,
  listComponent,
]).name;

export * from "./employee.service";
