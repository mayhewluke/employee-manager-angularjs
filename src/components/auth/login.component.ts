import angular, { IComponentOptions } from "angular";

import { logIn } from "./auth.service";

const config: IComponentOptions = {
  bindings: {
    email: "<",
    password: "<",
  },
  controller: class LoginFormController {
    public email!: string;
    public password!: string;
    public hasError: boolean = false;

    public submit(email: string, password: string) {
      logIn(email, password)
        .then(() => {
          this.hasError = false;
        })
        .catch(() => {
          this.hasError = true;
        });
    }
  },
  // TODO separate out template
  // TODO add spinner when request is in progress
  template: `
  <form ng-submit="$ctrl.submit($ctrl.email, $ctrl.password)">
    <input name="email" type="text" ng-model="$ctrl.email" required>
    <input name="password" type="password" ng-model="$ctrl.password" required>
    <button type="submit">Log in</button>
    <p ng-if="$ctrl.hasError" class="error">Something went wrong!</p>
  </form>
  `,
};

const name = angular.module("loginComponent", []).component("login", config)
  .name;

export default name;
