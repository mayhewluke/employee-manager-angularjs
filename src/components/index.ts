import angular from "angular";

import { authModule } from "./auth";

export const componentsModule = angular.module("app.components", [authModule])
  .name;
