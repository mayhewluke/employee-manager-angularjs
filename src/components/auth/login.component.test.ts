jest.mock("components/auth/auth.service");

import angular, { ICompileService, IQService } from "angular";
import "angular-mocks";

import { logIn } from "components/auth";

import loginComponent from "./login.component";

describe("loginComponent", () => {
  let render: (email?: string) => JQLite;
  let scope: any;
  let $q: IQService;
  beforeEach(angular.mock.module(loginComponent));
  beforeEach(
    angular.mock.inject(
      // tslint:disable-next-line:variable-name
      ($rootScope, $compile: ICompileService, _$q_: IQService) => {
        $q = _$q_;
        scope = $rootScope.$new();
        render = email => {
          scope.email = email;
          const ele = $compile(`<login email="email"></login>`)(scope);
          scope.$digest();
          return ele;
        };
        (logIn as any).mockImplementation(() => $q(() => null));
      },
    ),
  );

  it("does not show an error when first rendered", () => {
    const element = render();

    expect(element.find(".error").length).toBe(0);
  });

  describe("when an email is passed in", () => {
    it("prepopulates the email input", () => {
      const email = "user@domain.com";
      const element = render(email);

      expect(element.find("input[name=email]").val()).toEqual(email);
    });
  });

  describe("when no email is passed in", () => {
    it("leaves the email input blank", () => {
      const element = render();

      expect(element.find("input[name=email]").val()).toEqual("");
    });
  });

  describe("when submitted", () => {
    it("attempts to log in", () => {
      const email = "user@example.com";
      const password = "somethingsecret";
      const element = render();

      element
        .find("input[name='email']")
        .val(email)
        .triggerHandler("input");
      element
        .find("input[name='password']")
        .val(password)
        .triggerHandler("input");

      element.find("form").triggerHandler("submit");

      expect(logIn).toHaveBeenCalledTimes(1);
      expect(logIn).toHaveBeenCalledWith(email, password);
    });

    describe("when logging in fails", () => {
      let element: JQLite;
      beforeEach(() => {
        (logIn as any).mockImplementation(() => $q.reject());
        element = render();

        element.find("form").triggerHandler("submit");
        scope.$digest();
      });

      it("shows an error message", () => {
        expect(element.find(".error")[0].outerHTML).toMatchSnapshot();
      });

      describe("when resubmitted", () => {
        it("clears the error message", () => {
          (logIn as any).mockImplementation(() => $q.resolve());

          element.find("form").triggerHandler("submit");
          scope.$digest();

          expect(element.find(".error").length).toBe(0);
        });
      });
    });
  });
});
