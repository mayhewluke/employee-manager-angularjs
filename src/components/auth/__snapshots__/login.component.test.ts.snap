// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginComponent when submitted when logging in fails shows an error message 1`] = `"<p ng-if=\\"$ctrl.hasError\\" class=\\"error ng-scope\\">Something went wrong!</p>"`;
